### PR TITLE
[mysql] Refactor split assigner and simplify chunk splitting

### PR DESCRIPTION
**Pull request by wuchong**
1) This commit refactors MysqlSplitAssigner into an interface with different implementation, e.g. snapshot, binlog, hybrid.
2) This also simplifies chunk splitting logic with a ChunkSplitter util.
3) We also move global order guarantee of snapshot and binlog records from SourceReader to SplitAssigner, as it naturally a waiting condition for switching from snapshot splits to binlog splits.

### CODE REVIEW

File: flink-connector-mysql-cdc/src/main/java/com/alibaba/ververica/cdc/connectors/mysql/debezium/DebeziumUtils.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+            throw new FlinkRuntimeException("Failed to open MySQL connection", e);
+        }
+
+        return jdbc;
+    }
+
+    /** Closes the given {@link MySqlConnection}. */
+    public static void closeMySqlConnection(MySqlConnection jdbc) {
+        try {
+            jdbc.close();
```
Comment (leonardBang) [inline flink-connector-mysql-cdc/src/main/java/com/alibaba/ververica/cdc/connectors/mysql/debezium/DebeziumUtils.java:69]: please check null

File: flink-connector-mysql-cdc/src/main/java/com/alibaba/ververica/cdc/connectors/mysql/schema/MySqlTypeUtils.java [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    private static final String VARCHAR = "VARCHAR";
+    private static final String TINYTEXT = "TINYTEXT";
+    private static final String MEDIUMTEXT = "MEDIUMTEXT";
+    private static final String TEXT = "TEXT";
+    private static final String LONGTEXT = "LONGTEXT";
+    private static final String DATE = "DATE";
+    private static final String TIME = "TIME";
+    private static final String DATETIME = "DATETIME";
+    private static final String TIMESTAMP = "TIMESTAMP";
+    private static final String YEAR = "YEAR";
```
Comment (leonardBang) [inline flink-connector-mysql-cdc/src/main/java/com/alibaba/ververica/cdc/connectors/mysql/schema/MySqlTypeUtils.java:57]: Does `FLOAT_UNSIGNED` and `YEAR` Flink SQL supports them,  I think we should aligned with `RowDataDebeziumDeserializeSchema`

File: flink-connector-mysql-cdc/src/main/java/com/alibaba/ververica/cdc/connectors/mysql/source/assigners/ChunkSplitter.java [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+        final Object[] minMaxOfSplitColumn = queryMinMax(jdbc, tableId, splitColumnName);
+        final Object min = minMaxOfSplitColumn[0];
+        final Object max = minMaxOfSplitColumn[1];
+        if (min == null || max == null || min.equals(max)) {
+            // empty table, or only one row, return full table scan as a chunk
+            return Collections.singletonList(ChunkRange.all());
+        }
+
+        final List<ChunkRange> chunks;
+        if (splitColumnUniformlyDistributed(splitColumn)) {
```
Comment (leonardBang) [inline flink-connector-mysql-cdc/src/main/java/com/alibaba/ververica/cdc/connectors/mysql/source/assigners/ChunkSplitter.java:132]: we can keep notes `evenly` same with method signature `Uniformly`

File: flink-connector-mysql-cdc/src/test/java/com/alibaba/ververica/cdc/connectors/mysql/source/split/MySqlSplitSerializerTest.java [hunk 4]
```diff
@@ -96,7 +95,7 @@ public void testBinlogSplit() throws Exception {
 
         final MySqlSplit split =
                 new MySqlBinlogSplit(
-                        BINLOG_SPLIT_ID,
+                        "binlog-split",
```
Comment (leonardBang) [inline flink-connector-mysql-cdc/src/test/java/com/alibaba/ververica/cdc/connectors/mysql/source/split/MySqlSplitSerializerTest.java:98]: nit: use `BINLOG_SPLIT_ID`

File: flink-connector-mysql-cdc/src/test/java/com/alibaba/ververica/cdc/connectors/mysql/table/MySqlConnectorITCase.java [hunk 5]
```diff
@@ -115,9 +115,9 @@ public void testConsumingAllEvents() throws Exception {
                                 + " 'database-name' = '%s',"
                                 + " 'table-name' = '%s',"
                                 + " 'debezium.internal.implementation' = '%s',"
-                                + " 'scan.snapshot.parallel-read' = '%s',"
+                                + " 'scan.incremental.snapshot.enabled' = '%s',"
                                 + " 'server-id' = '%s',"
```
Comment (leonardBang) [inline flink-connector-mysql-cdc/src/test/java/com/alibaba/ververica/cdc/connectors/mysql/table/MySqlConnectorITCase.java:119]: Also rename `parallel-read` to `incrementalSnapsshot` ?

File: flink-connector-mysql-cdc/src/main/java/com/alibaba/ververica/cdc/connectors/mysql/source/assigners/ChunkSplitter.java [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+    private final int chunkSize;
+
+    public ChunkSplitter(MySqlConnection jdbc, MySqlSchema mySqlSchema, int chunkSize) {
+        this.jdbc = jdbc;
+        this.mySqlSchema = mySqlSchema;
+        this.chunkSize = chunkSize;
+    }
+
+    /** Generates all snapshot splits (chunks) for the give table path. */
+    public Collection<MySqlSnapshotSplit> enumerateSplits(TableId tableId) {
```
Comment (leonardBang) [inline flink-connector-mysql-cdc/src/main/java/com/alibaba/ververica/cdc/connectors/mysql/source/assigners/ChunkSplitter.java:74]: how about rename to `generateSplits`?

**Review by leonardBang — COMMENTED** [review]
Thanks @wuchong for the great refactor, I left minor comments

File: flink-connector-mysql-cdc/src/main/java/com/alibaba/ververica/cdc/connectors/mysql/schema/MySqlTypeUtils.java (hunk 2 above)
Comment (wuchong) [reply flink-connector-mysql-cdc/src/main/java/com/alibaba/ververica/cdc/connectors/mysql/schema/MySqlTypeUtils.java:57]: We didn't say we support `FLOAT_UNSIGNED` and `YEAR` in docs https://github.com/ververica/flink-cdc-connectors/wiki/MySQL-CDC-Connector#data-type-mapping

File: flink-connector-mysql-cdc/src/test/java/com/alibaba/ververica/cdc/connectors/mysql/source/split/MySqlSplitSerializerTest.java (hunk 4 above)
Comment (wuchong) [reply flink-connector-mysql-cdc/src/test/java/com/alibaba/ververica/cdc/connectors/mysql/source/split/MySqlSplitSerializerTest.java:98]: I just don't want to make `BINLOG_SPLIT_ID` to be public, and we should be able to use arbitrary split id here.

**Review by leonardBang — APPROVED** [review]
LGTM

**Pull request merged**